docs/adrs/004-SplitAndAggregate.md

```diff
@@ -0,0 +1,198 @@
+* `-` Creating a new queue requires balancing between how many active KBP nodes there vs how many nodes are available for other NJS jobs.
+* `-` Capacity is still wasted, as some KBP jobs sit around waiting for other jobs to run
+
+
+### Build KBP Lightweight Version + KBP Queue
+
+
+#### Design of new verison
+* All apps must be modified to use the new KBP lightweight version, which will: 
+* Can either modify KBP, or create a new tool/package to use instead of KBP
+
+1) Launch a management job called the *Job Manager* that sits in the KBP Queue, alongside other KBP jobs. Other jobs are launched in the NJS queue.
+2) Launch the *Setup Job* which will
+ * Use the *User Parameters* and/or
+ * Optionally Download the Data from the *User Parameters*  to figure out *Job Manager* parameters
+ * Use the results of information gathered from the initial download and or *User Parameters*
+ * Generate final parameters to be sent to the *Job Manager* to launch *Fan Out* jobs
+3) The *Job Manager* now has enough parameters to setup *Fan Out* Jobs, and will send out jobs and wait for them (and possibly retry them upon failure)
+4) *Fan Out* jobs download data and perform calculations, save them back to the system, and return references to the saved objects
+5) The *Job Manager* optionally launches a *Group/Reduce* job based on User Parameters and or the results of *Fan Out* Jobs
+6) The *Group/Reduce* job downloads objects from the system, and creates a set or other grouping, and then saves the set back to the system
+7) The *Job Manager*  sends the results back to a *Report* Job, which downloads the final data and uploads a report based on that data
+8) The *Job Manager* returns the reference to the results of the *Report Job*
+
+Pros/Cons
+* `-` Addresses the deadlocking issue, UI still broken for regular runs and batch runs if we re-use KBP
+* `-` Have to rewrite all apps that use KBP to use this new paradigm
+* `-` Will have to upload data and download data more times as no computation is done in the *Job Manager*
+* `+` All KBP jobs can run on a small subset of machines, deadlock issue is fixed
+* `+` No changes to ee2 required
+
+
+ 
+### Modify Apps to do only local submission by remove KBP, and moving the job 
+* `+` Simple solutions, quick turnarounds, fixes deadlock issue, fixes UI issues
+* `-` We have a limited number of larger resources machines
+* `-` Continued dependency on deprecated KBP tools
+* `-` App runs may take longer since fewer resources may be available to the app run
+
+### Deprecate kbparallels, and write new ee2 endpoints to entirely handle split and aggregate
+* `+` No longer uses an app, No longer uses a slot in the queue
+* `-` All jobs would have to change to stop using KBP in favor of these via the ee2 client,
+* `+` Can manage jobs more closely: have tables, more control over lifecycle, resource requests, canceling subjobs
+* `-` Job monitoring thread would have to run in ee2
```

Comment:
I don't think this is true - EE2 would just have to check if all jobs are done when any PIP job is done. You don't need a thread for that.

Reply:
That sounds right

Reply:
:+1: